### PR TITLE
[timeseries] Add native support for missing values

### LINKED ISSUE #3886: [timeseries] Add first-class support for missing values
Currently, `TimeSeriesPredictor` deals with missing values in the data by first imputing them via forward-/backward-filling, and then training all models as if there are no missing values. This strategy may lead to poor accuracy on datasets with a large portion of missing values since training data will include regions with constant values arising from ffilling. 

A better alternative is to keep the missing values represented by `nan` in the data, and let models handle it. This requires following modifications to the code:
- [x] Update metric implementations to handle missing values in target (#3966)
- [ ] Update preprocessing logic in `TimeSeriesPredictor._check_and_prepare_data_frame`
- [ ] Make sure that all models can handle missing values. This means, all models can train normally and produce forecasts with no NaN values, even if training data contains NaNs.
  - [ ] GluonTS models (DeepAR, TFT, PatchTST, DLinear)
  - [ ] StatsForecast models (AutoETS, AutoARIMA, Theta, intermittent demand models)
  - [ ] Baseline models (Naive, SeasonalNaive, Average, SeasonalAverage, Zero)
  - [ ] MLForecast models (DirectTabular, RecursiveTabular)

### PR DESCRIPTION
**Pull request by shchur**
*Issue #, if available:* fixes #3886 

*Description of changes:*
- Instead of imputing all missing values in `target` column in `TimeSeriesPredictor._check_and_prepare_data_frame`, we let each model use its own logic for handling missing values.
  - GluonTS models (DeepAR, TFT, PatchTST, DLinear) + some local models (Average, SeasonalAverage, NPTS, Naive, SeasonalNaive) handle the missing values natively
  - Other local models (AutoETS, AutoCES, AutoARIMA, Theta, intermittent models) perform imputation first
  - MLForecast models use a mix of two strategies: missing values are imputed, but rows that originally contained NaN values are not used for training
- Model properties (e.g., whether it can handle missing values) are stored using the `_get_tags()` mechanism
- `TimeSeriesPredictor` now removes time series consisting of only NaN values from `train_data` during `fit()`
- Missing values in covariates are still always imputed inside `TimeSeriesFeatureGenerator`
- Add missing values to `DUMMY_TS_DATAFRAME` used in the tests to ensure that NaN support works in all scenarios

*To do:*
- [x] Add tests
- [x] Add missing values support to MLForecast models
- [x] Benchmark the new NaN handling strategy on datasets with missing values

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by yinweisu** [comment]
| Previous CI Run | Current CI Run |
| :---: | :---: |

**Comment by yinweisu** [comment]
| Previous CI Run | Current CI Run |
| :---: | :---: |

**Comment by yinweisu** [comment]
| Previous CI Run | Current CI Run |
| :---: | :---: |
| huggingface-hub==0.21.4 | huggingface-hub==0.22.0 |
| lightning-utilities==0.11.0 | lightning-utilities==0.11.1 |
| filelock==3.13.1 | filelock==3.13.2 |
| huggingface-hub==0.21.4 | huggingface-hub==0.22.0 |
| lightning-utilities==0.11.0 | lightning-utilities==0.11.1 |
| filelock==3.13.1 | filelock==3.13.2 |

File: timeseries/src/autogluon/timeseries/models/autogluon_tabular/mlforecast.py [hunk 1]
```diff
@@ -85,6 +85,21 @@ def __init__(
...
         self._residuals_std_per_item: Optional[pd.Series] = None
         self._avg_residuals_std: Optional[float] = None
+        self._train_target_median: Optional[float] = None
+
+    def preprocess(self, data: TimeSeriesDataFrame, is_train: bool = False, **kwargs) -> Any:
+        if is_train:
+            # All-NaN series are removed; partially-NaN series in train_data are handled inside _generate_train_val_dfs
+            all_nan_items = data.item_ids[data[self.target].isna().groupby(ITEMID, sort=False).all()]
+            if len(all_nan_items):
+                data = data.query("item_id not in @all_nan_items")
```
Comment (canerturkmen) [inline timeseries/src/autogluon/timeseries/models/autogluon_tabular/mlforecast.py:95]: Just to check, this should never happen when called through the predictor?

File: timeseries/tests/unittests/models/test_models.py [hunk 2]
```diff
@@ -475,3 +481,91 @@ def test_given_searcher_when_ray_backend_used_in_hpo_then_correct_searcher_used(
...
+        train_data=data,
+        val_data=None if isinstance(model, MultiWindowBacktestingModel) else data,
+    )
+    predictions = model.predict(data)
+    assert not predictions.isna().any(axis=None) and all(predictions.item_ids == data.item_ids)
+
+
+@pytest.mark.parametrize("model_class", TESTABLE_MODELS)
+def test_when_fit_and_predict_called_then_train_val_and_test_data_is_preprocessed(temp_model_path, model_class):
+    train_data = DUMMY_TS_DATAFRAME_WITH_MISSING.copy()
```
Comment (canerturkmen) [inline timeseries/tests/unittests/models/test_models.py:506]: Just to check: We never modify the user-provided dataframe inplace, right?

**Review by canerturkmen — COMMENTED** [review]
LGTM overall thanks a lot for this! Dropped a few minor comments and questions.

File: timeseries/src/autogluon/timeseries/utils/features.py [hunk 3]
```diff
@@ -183,7 +194,16 @@ def transform(self, data: TimeSeriesDataFrame, data_frame_name: str = "data") ->
         else:
             static_features = None
 
-        return TimeSeriesDataFrame(pd.concat(dfs, axis=1), static_features=static_features)
+        ts_df = TimeSeriesDataFrame(pd.concat(dfs, axis=1), static_features=static_features)
+
+        covariates_names = self.covariate_metadata.covariates
+        if len(covariates_names) > 0:
+            ts_df[covariates_names] = ts_df[covariates_names].fill_missing_values()
+            # If some series consist completely of NaNs, fill with median
```
[the remark this answers is not in the document]
Comment (shchur) [reply timeseries/src/autogluon/timeseries/utils/features.py:202]: The first line 
```python
ts_df[covariates_names] = ts_df[covariates_names].fill_missing_values()
```
does ffill + bfill for all covariates (per item). If there are still missing values left after that, it means that for some time series that covariate consists completely of NaNs. In this case, we use the median of the covariate in the dataset for imputation.

Thinking about it, we probably need to store the median covariate values of the training data during fit and use those for imputation. This would make more sense in case user tries to predict items one by one (and for some items the covariate is all-nan)

File: timeseries/src/autogluon/timeseries/models/abstract/abstract_timeseries_model.py [hunk 4]
```diff
@@ -235,7 +237,10 @@ def fit(self, **kwargs) -> "AbstractTimeSeriesModel":
         model: AbstractTimeSeriesModel
             The fitted model object
         """
-        return super().fit(**kwargs)
+        train_data = self.preprocess(train_data, is_train=True)
+        if self._get_tags()["can_use_val_data"] and val_data is not None:
```
[the remark this answers is not in the document]
Comment (canerturkmen) [reply timeseries/src/autogluon/timeseries/models/abstract/abstract_timeseries_model.py:241]: Got it. Thanks!

File: timeseries/src/autogluon/timeseries/models/local/npts.py [hunk 5]
```diff
@@ -88,3 +88,6 @@ def _predict_with_local_model(
         for q in self.quantile_levels:
             forecast_dict[str(q)] = forecast.quantile(q)
         return pd.DataFrame(forecast_dict)
+
+    def _more_tags(self) -> dict:
+        return {"allow_nan": True}
```
[the remark this answers is not in the document]
Comment (canerturkmen) [reply timeseries/src/autogluon/timeseries/models/local/npts.py:93]: To better articulate my question: if the target start position is not nan but the sampled trajectory does contain a NaN, would the forecast have NaNs?

File: timeseries/src/autogluon/timeseries/utils/features.py [hunk 6]
```diff
@@ -17,16 +17,27 @@
 logger = logging.getLogger(__name__)
 
 
-@dataclass
-class CovariateMetadata:
+class CovariateMetadata(BaseModel):
```
[the remark this answers is not in the document]
Comment (canerturkmen) [reply timeseries/src/autogluon/timeseries/utils/features.py:20]: It can still complicate things due to dependency version conflicts or just one more thing to maintain. I would rather not introduce a dependency for this sort of functionality, but your call.

File: timeseries/src/autogluon/timeseries/utils/features.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (canerturkmen) [reply timeseries/src/autogluon/timeseries/utils/features.py:202]: That's quite complicated logic. If we can get away with filling zeros for this edge case for example, I'm OK with that (with a loud warning as well) just to save complexity.

File: timeseries/src/autogluon/timeseries/utils/features.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (shchur) [reply timeseries/src/autogluon/timeseries/utils/features.py:20]: Fair enough, I guess this can be handled by a static type checker without introducing a dependency

**Comment by yinweisu** [comment]
| Previous CI Run | Current CI Run |
| :---: | :---: |
| flatbuffers==24.3.7 | flatbuffers==24.3.25 |
| flatbuffers==24.3.7 | flatbuffers==24.3.25 |

**Comment by yinweisu** [comment]
| Previous CI Run | Current CI Run |
| :---: | :---: |

File: timeseries/tests/unittests/test_features.py [hunk 7]
```diff
@@ -135,3 +135,35 @@ def test_when_bool_columns_provided_then_they_are_converted_to_cat():
...
+
+    data_transformed = feat_generator.fit_transform(data)
+    assert not data_transformed[feat_generator.covariate_metadata.covariates].isna().any(axis=None)
+    assert data_transformed["target"].isna().any()
+
+    known_covariates_transformed = feat_generator.transform_future_known_covariates(known_covariates)
+    if known_covariates_names == []:
+        assert known_covariates_transformed is None
+    else:
+        assert not known_covariates_transformed[known_covariates_names].isna().any(axis=None)
```
Comment (canerturkmen) [inline timeseries/tests/unittests/test_features.py:158]: This looks great and I'll probably use it for feature importance.

Should we also test the filling logic, at least for the 'median' and 'mode' scenarios?

**Review by canerturkmen — APPROVED** [review]
LGTM! Only one comment.

**Comment by yinweisu** [comment]
| Previous CI Run | Current CI Run |
| :---: | :---: |

**Comment by yinweisu** [comment]
| Previous CI Run | Current CI Run |
| :---: | :---: |

**Comment by yinweisu** [comment]
| Previous CI Run | Current CI Run |
| :---: | :---: |

**Review by shchur — COMMENTED** [review]
Thanks @canerturkmen, I've added the tests for mode/median imputation + added missing values to `DUMMY_TS_DATAFRAME` so that we consider more settings with NaNs

**Comment by shchur** [comment]
@canerturkmen just finished some benchmarking on 12 datasets with missing values / 3 folds each: This PR branch has 62.5% win rate vs. current `master` branch, so I guess we are good to merge once the problems with tests are resolved.

**Comment by yinweisu** [comment]
| Previous CI Run | Current CI Run |
| :---: | :---: |

File: timeseries/tests/unittests/models/test_local.py [hunk 8]
```diff
@@ -86,25 +86,6 @@ def test_when_local_model_saved_then_local_model_args_are_saved(model_class, hyp
     assert dict_equal_primitive(model._local_model_args, loaded_model._local_model_args)
 
 
-@pytest.mark.parametrize("model_class", TESTABLE_MODELS)
```
Comment (shchur) [inline timeseries/tests/unittests/models/test_local.py:89]: We already have the same test here https://github.com/autogluon/autogluon/blob/master/timeseries/tests/unittests/models/test_models.py#L269.

**Comment by yinweisu** [comment]
| Previous CI Run | Current CI Run |
| :---: | :---: |

**Comment by yinweisu** [comment]
| Previous CI Run | Current CI Run |
| :---: | :---: |

**Review by canerturkmen — APPROVED** [review]
LGTM! 🚀
